orthanc/orthanc-raw/plugin/pixl.py

```diff
@@ -72,6 +79,53 @@ def should_auto_route():
     return os.environ.get("ORTHANC_AUTOROUTE_RAW_TO_ANON", "false").lower() == "true"
 
 
+def write_dataset_to_bytes(dataset: Dataset) -> bytes:
+    """
+    Write pydicom DICOM dataset to byte array
+
+    Original from:
+    https://pydicom.github.io/pydicom/stable/auto_examples/memory_dataset.html
+    """
+    with BytesIO() as buffer:
+        dcmwrite(buffer, dataset)
+        buffer.seek(0)
+        return buffer.read()
+
+
+def modify_dicom_tags(receivedDicom: bytes, origin: str) -> Any:
+    """
+    A new incoming DICOM file needs to have the project name private tag added here, so
+    that the API will later allow us to edit it.
+    However, we don't know its correct value at this point, so just create it with an obvious
+    placeholder value.
+    """
+    if origin != orthanc.InstanceOrigin.DICOM_PROTOCOL:
+        # don't keep resetting the tag values if this was triggered by an API call!
+        print("modify_dicom_tags - doing nothing as change triggered by API")  # noqa: T201
+        return orthanc.ReceivedInstanceAction.KEEP_AS_IS, None
+    dataset = dcmread(BytesIO(receivedDicom))
+    private_creator_name = "UCLH PIXL"
+    # See the orthanc.json config file for where this tag is given a nickname
+    private_tag_offset = 0x01
+    # LO = Long string max 64, LT = long text max 10240, support paragraphs etc
+    # https://dicom.nema.org/medical/dicom/current/output/chtml/part05/sect_6.2.html
+    vr = "LO"
+    unknown_value = "__pixl_unknown_value__"
+    group_id = 0x000B
+    # The private block is the first free block >= 0x10. Other parts of the code assume
+    # it is == 0x10 though :/
+    # https://dicom.nema.org/dicom/2013/output/chtml/part05/sect_7.8.html
```

Comment:
@jstutters Did you say there was some solution for this? Were we going to just pick a higher value and hope it's free or something like that?

Reply:
Yes, sadly the best we've got at the moment is to pick something high and unlikely to be selected for something else.  pydicom can search by name but the Orthanc modify method requires that the hex value be associated with a name in advance.  It should only be necessary to change the group ID.

Reply:
I assume we're talking about the private block ID here? Ie. the BB in GGGG,BBOO.

This comment in the configuration file in the source code is interesting:
https://orthanc.uclouvain.be/hg/orthanc/file/Orthanc-1.12.3/OrthancServer/Resources/Configuration.json#l689

> Note: For private tags, you should only declare the lower 8 bits
> of the element since the higher 8 bits may vary from one file to
> the other.

I've tried setting the dictionary entry to "000b,01" and it doesn't immediately work, but this might be worth persuing as it's very much supposed to.

Reply:
Interestingly, the orthanc server is recognising the value as expected, even if it ends up in 0x1101. It's the API that's the problem.
![image](https://github.com/UCLH-Foundry/PIXL/assets/5707853/00b71ea4-7652-4836-90f0-acfba8870c30)

Reply:
pydicom is really not geared to the user explicitly chosing their block number, as far as I can tell. It's not supposed to matter but orthanc doesn't handle it right. Can we try using a higher group number (eg. 0x000d) and hoping we get block 0x10?

Reply:
I have moved it to group 0x000D and will mark this as resolved.